haystack/document_stores/elasticsearch_base.py

```diff
@@ -0,0 +1,1696 @@
+from typing import List, Optional, Union, Dict, Any, Generator
+
+import json
+import logging
+import time
+from copy import deepcopy
+from string import Template
+
+import numpy as np
+from scipy.special import expit
+from tqdm.auto import tqdm
+from pydantic.error_wrappers import ValidationError
+
+from haystack.document_stores import KeywordDocumentStore
+from haystack.schema import Document, Label
+from haystack.document_stores.base import get_batches_from_generator
+from haystack.document_stores.filter_utils import LogicalFilterClause
+from haystack.errors import DocumentStoreError, HaystackError
+from haystack.nodes.retriever import DenseRetriever
+
+
+logger = logging.getLogger(__name__)
+
+
+def prepare_hosts(host, port):
+    """
+    Create a list of host(s) + port(s) to allow direct client connections to multiple elasticsearch nodes,
+    in the format expected by the client.
+    """
+    if isinstance(host, list):
+        if isinstance(port, list):
+            if not len(port) == len(host):
+                raise ValueError("Length of list `host` must match length of list `port`")
+            hosts = [{"host": h, "port": p} for h, p in zip(host, port)]
+        else:
+            hosts = [{"host": h, "port": port} for h in host]
+    else:
+        hosts = [{"host": host, "port": port}]
+    return hosts
+
+
+class BaseElasticsearchDocumentStore(KeywordDocumentStore):
```

Comment:
As this class is the base class for both `ElasticsearchDocumentStore` and `OpenSearchDocumentStore`, I was wondering whether `BaseElasticsearchDocumentStore` is the best name. However, I have no alternative suggestion so I would be happy with keeping it as is. I saw that some of the logging/exception messages in this class only say `Elasticsearch` where I think they should say `Elasticsearch/OpenSearch` instead.

Reply:
I agree, we had a conversation with @ZanSara when we first added the base class that was inconclusive, we couldn't come up with a better name either. I'll take some time to think again about a better name tho.